File: packages/domain/src/rule.ts
export type TriggerKind =
    | { kind: 'on_item_create' }
    | { kind: 'on_item_update' }
    | { kind: 'on_item_delete' }
    | { kind: 'on_comment_add' }
    | { kind: 'on_type_assign';    type_name: string | null }
    | { kind: 'on_type_unassign';  type_name: string | null }
    | { kind: 'on_attribute_set';  attribute_key: string | null }
    | { kind: 'cron';              expression: string }
    | { kind: 'interval';          seconds: number }
    | { kind: 'manual' };

export type TriggerKindTag = TriggerKind['kind'];

export const TRIGGER_KIND_LABELS: Record<TriggerKindTag, string> = {
    on_item_create:   'On Item Created',
    on_item_update:   'On Item Updated',
    on_item_delete:   'On Item Deleted',
    on_comment_add:   'On Comment Added',
    on_type_assign:   'On Type Assigned',
    on_type_unassign: 'On Type Unassigned',
    on_attribute_set: 'On Attribute Set',
    cron:             'Scheduled (Cron)',
    interval:         'Scheduled (Interval)',
    manual:           'Manual',
};

export class Rule {
    constructor(
        public readonly rule_id:     string,
        public readonly account_id:  string,
        public readonly name:        string,
        public readonly description: string,
        public readonly trigger:     TriggerKind,
        public readonly script:      string,
        public readonly enabled:     boolean,
        public readonly created_at:  string,
        public readonly last_run_at: string | null,
        public readonly last_error:  string | null,
        public readonly last_output: string | null,
    ) {}

    static fromDto(dto: RuleDto): Rule {
        return new Rule(
            dto.rule_id,
            dto.account_id,
            dto.name,
            dto.description,
            dto.trigger,
            dto.script,
            dto.enabled,
            dto.created_at,
            dto.last_run_at,
            dto.last_error,
            dto.last_output,
        );
    }
}

export interface RuleDto {
    rule_id:     string;
    account_id:  string;
    name:        string;
    description: string;
    trigger:     TriggerKind;
    script:      string;
    enabled:     boolean;
    created_at:  string;
    last_run_at: string | null;
    last_error:  string | null;
    last_output: string | null;
}

export interface AddRuleDto {
    name:         string;
    description?: string;
    trigger:      TriggerKind;
    script:       string;
    enabled?:     boolean;
}

export interface UpdateRuleDto {
    name?:        string;
    description?: string;
    trigger?:     TriggerKind;
    script?:      string;
    enabled?:     boolean;
}

export interface RuleRunResult {
    rule_id:     string;
    success:     boolean;
    output:      string | null;
    error:       string | null;
    duration_ms: number;
}

File: crates/zealot-domain/src/rule.rs
use crate::common::id::Id;
use chrono::NaiveDateTime;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone)]
pub struct Rule {
    pub rule_id: Id,
    pub account_id: Id,
    pub name: String,
    pub description: String,
    pub trigger: TriggerKind,
    pub script: String,
    pub enabled: bool,
    pub created_at: NaiveDateTime,
    pub last_run_at: Option<NaiveDateTime>,
    pub last_error: Option<String>,
    pub last_output: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum TriggerKind {
    OnItemCreate,
    OnItemUpdate,
    OnItemDelete,
    OnCommentAdd,
    OnTypeAssign { type_name: Option<String> },
    OnTypeUnassign { type_name: Option<String> },
    OnAttributeSet { attribute_key: Option<String> },
    Cron { expression: String },
    Interval { seconds: u64 },
    Manual,
}

impl TriggerKind {
    pub fn kind_str(&self) -> &'static str {
        match self {
            Self::OnItemCreate => "on_item_create",
            Self::OnItemUpdate => "on_item_update",
            Self::OnItemDelete => "on_item_delete",
            Self::OnCommentAdd => "on_comment_add",
            Self::OnTypeAssign { .. } => "on_type_assign",
            Self::OnTypeUnassign { .. } => "on_type_unassign",
            Self::OnAttributeSet { .. } => "on_attribute_set",
            Self::Cron { .. } => "cron",
            Self::Interval { .. } => "interval",
            Self::Manual => "manual",
        }
    }

    pub fn is_event_triggered(&self) -> bool {
        matches!(
            self,
            Self::OnItemCreate
                | Self::OnItemUpdate
                | Self::OnItemDelete
                | Self::OnCommentAdd
                | Self::OnTypeAssign { .. }
                | Self::OnTypeUnassign { .. }
                | Self::OnAttributeSet { .. }
        )
    }

    pub fn is_scheduled(&self) -> bool {
        matches!(self, Self::Cron { .. } | Self::Interval { .. })
    }
}

/// Sent to the client.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RuleDto {
    pub rule_id: i64,
    pub account_id: i64,
    pub name: String,
    pub description: String,
    pub trigger: TriggerKind,
    pub script: String,
    pub enabled: bool,
    pub created_at: NaiveDateTime,
    pub last_run_at: Option<NaiveDateTime>,
    pub last_error: Option<String>,
    pub last_output: Option<String>,
}

impl From<&Rule> for RuleDto {
    fn from(r: &Rule) -> Self {
        Self {
            rule_id: r.rule_id.into(),
            account_id: r.account_id.into(),
            name: r.name.clone(),
            description: r.description.clone(),
            trigger: r.trigger.clone(),
            script: r.script.clone(),
            enabled: r.enabled,
            created_at: r.created_at,
            last_run_at: r.last_run_at,
            last_error: r.last_error.clone(),
            last_output: r.last_output.clone(),
        }
    }
}

/// Received when creating a rule.
#[derive(Debug, Clone, Deserialize)]
pub struct AddRuleDto {
    pub name: String,
    pub description: Option<String>,
    pub trigger: TriggerKind,
    pub script: String,
    pub enabled: Option<bool>,
}

/// Received when updating a rule.
#[derive(Debug, Clone, Deserialize)]
pub struct UpdateRuleDto {
    pub name: Option<String>,
    pub description: Option<String>,
    pub trigger: Option<TriggerKind>,
    pub script: Option<String>,
    pub enabled: Option<bool>,
}
